### PR TITLE
Flex

### PR DESCRIPTION
**Pull request by rickycpadilla**
## Description
This PR adds the methods necessary for Passage Flex Android developers:

### PassageFlex methods:
```kotlin
PassageFlex.Passkey.register(transactionId, activity)
PassageFlex.Passkey.authenticate(transactionId, activity)
```

## Example usage
```kotlin
try {
  // Exchange a user identifier for a transaction id on your app's server.
  
  // Prompt your user to create and register a passkey:
  val passageNonce = PassageFlex.Passkey.register(transactionId, activity)
  
  // Send `passageNonce` to your app server to finish authentication and authorization.
} catch (e: RegisterCancellationException) {
  // User canceled request, do nothing.
} catch {
  // Show error UI
}
```

## PR does NOT include:
* Physical Security Key compatibility
* README
* Testing

### Duplicate code
This new package contains code from the standard Passage Android package, including generated code. In the near future we should find a better way to share code across the two packages.

### CODE REVIEW

File: passageflex/src/main/java/id/passage/passageflex/AuthenticateException.kt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            val message = e.message ?: e.toString()
+            return when (e) {
+                is GetCredentialException -> convertCreateCredentialException(e)
+                is ClientException -> convertClientException(e)
+                is ServerException -> AuthenticateServerException(message)
+                else -> AuthenticateException(message)
+            }
+        }
+
+        private fun convertClientException(e: ClientException): AuthenticateException {
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/AuthenticateException.kt:39]: I like these, will cross-reference with some of the existing passageErrors

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import id.passage.android.passageflex.api.AuthenticateAPI
+import id.passage.android.passageflex.api.RegisterAPI
+import id.passage.android.passageflex.model.AuthenticateWebAuthnFinishWithTransactionRequest
+import id.passage.android.passageflex.model.AuthenticateWebAuthnStartWithTransactionRequest
+import id.passage.android.passageflex.model.AuthenticatorAttachment
+import id.passage.android.passageflex.model.RegisterWebAuthnFinishWithTransactionRequest
+import id.passage.android.passageflex.model.RegisterWebAuthnStartWithTransactionRequest
+
+/**
+ *  Use `PassageFlex` to easily add passkeys to your existing authentication system on Apple devices.
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt:13]: Should this be "on Android devices"?

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ *
+ *  Find out more at: https://passage.1password.com
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt:15]: Also like referencing the website, stealing this!!

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ */
+public object PassageFlex {
+    /**
+     * The base struct for utilizing Apple's native passkey APIs and Passage Flex APIs together.
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt:19]: Again, I imagine this should be Android?

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+         * `transactionId` is provided, this method will attempt to show only passkeys associated
+         * with that user account.
+         *
+         * @param transactionId (Optional) The Passage transaction id provided by your app's server.
+         * @param activity Activity to surface the Credentials Manager prompt within.
+         * @return A single-use "nonce" from Passage server to be exchanged for an authentication
+         * token on your app's server.
+         * @throws AuthenticateException
+         */
+        public suspend fun authenticate(
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt:99]: TIL about `suspend`

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+         * @param transactionId The Passage transaction id provided by your app's server.
+         * @param activity Activity to surface the Credentials Manager prompt within.
+         * @param authenticatorAttachment (Optional) The type of authentication that will be used in
+         * this WebAuthN flow request. Defaults to `AuthenticatorAttachment.platform`. Use
+         * `AuthenticatorAttachment.cross-platform` for physical security key registration.
+         * @return A single-use "nonce" from Passage server to be exchanged for an authentication
+         * token on your app's server.
+         * @throws RegisterException
+         */
+        public suspend fun register(
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt:37]: Probably a nitpick but could/should the PassageFlex class be separated/broken down further? Would this help Android devs looking at the source code we expose in any way, or help make developing in Android flex any easier? In the JS PassageFlex I've kept the main class pretty barebones as an entrypoint and the passkey itself is a class, so are the authentication/register classes. It made sense in JS but not sure it makes sense here.

Just a thought and curious to hear yours!

File: passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+     * Registers a user passkey credential that can be used to authenticate the user to
+     * the app in the future.
+     *
+     * The execution potentially launches framework UI flows for a user to view their registration
+     * options, grant consent, etc.
+     *
+     * @param requestJson The privileged request in JSON format in the [standard webauthn web json](https://w3c.github.io/webauthn/#dictdef-publickeycredentialcreationoptionsjson).
+     * @throws CreateCredentialException If the request fails
+     */
+    public suspend fun createPasskey(
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt:41]: Another possibly heavily web FE biased (and my own personal bias haha) but I wonder what your thoughts are on separating out one util per file. There's nothing wrong with having a single util file and certainly nothing wrong with the way you have this laid out as the functions are in a similar vein, but I have seen these utils file sometimes turn into catch-alls over time that lead to code smells.

Again - my personal preference, a nitpick, and might not even make sense for mobile dev, I'm just curious!

File: passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+     */
+    internal fun getCredentialOptionsJson(challenge: CredentialAssertionChallenge?): String {
+        val credOptions =
+            challenge?.challenge?.publicKey
+                ?: throw CredentialParsingException(CHALLENGE_MISSING)
+        val moshi = Moshi.Builder().build()
+        val credOptionsAdapter = ProtocolCredentialAssertionPublicKeyJsonAdapter(moshi)
+        // Passage API bug: Login API frequently returns challenge with non-url-safe characters
+        // "+" and "/" that cause "Bad Base 64" exception to be thrown by the credential manager.
+        val modifiedChallenge = credOptions.challenge.replace('+', '-').replace('/', '_')
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt:123]: Hmm interesting. Is this something we should guard against in the JS sdk?

File: passageflex/src/main/java/id/passage/passageflex/Utils.kt [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package id.passage.passageflex
+
+import android.app.Activity
+
+internal class Utils {
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/Utils.kt:5]: So these utils are meant to apply to all of PassageFlex including non-passkey flex options in the future, correct?

**Review by jenn-macfarlane — APPROVED** [review]
This is looking great to me, just (what I think is) a couple typos where it should be Android vs. Apple, and a couple questions regarding codebase structure for kotlin codebases.

I did only go through the `/src` folder thoroughly - let me know if I should look into the auto-generated code or something outside of `/src` needs a look.

File: passageflex/src/main/java/id/passage/passageflex/RegisterException.kt [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                else -> RegisterException(message)
+            }
+        }
+    }
+}
+
+/**
+ * The user intentionally canceled the operation and chose not to register the credential.
+ */
+public class RegisterCancellationException(message: String) : RegisterException(message)
```
Comment (jenn-macfarlane) [inline passageflex/src/main/java/id/passage/passageflex/RegisterException.kt:81]: How do you determine this on mobile? I'd love to handle this on the JS side but not sure how atm

File: passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt (hunk 8 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt:123]: You should be good in JS :)

File: passageflex/src/main/java/id/passage/passageflex/RegisterException.kt (hunk 10 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/RegisterException.kt:81]: The Android Credential Manager explicitly throws a cancelation exception that I can then transform into this custom exception :)

Comment (jenn-macfarlane) [reply passageflex/src/main/java/id/passage/passageflex/RegisterException.kt:81]: Interesting. I wonder if I could intercept that on the web. The element for example just shows our standard failure UI when that happens but that would be up to the developer to build for JS. Will give it a thought!

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt (hunk 2 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt]: Yes, thank you!

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt (hunk 4 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt]: Yes, thank you!

File: passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt (hunk 6 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/PassageFlex.kt:31]: Thanks for the recommendation! I have broken it out further, matching the iOS structure.

File: passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt (hunk 7 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt:41]: Thanks for the input! I may break up this class in the future, but think its okay for now :)

File: passageflex/src/main/java/id/passage/passageflex/Utils.kt (hunk 9 above)
Comment (rickycpadilla) [reply passageflex/src/main/java/id/passage/passageflex/Utils.kt:5]: Yup!

**Pull request merged**